### ISSUE TITLE
[java] SwitchStmtsShouldHaveDefault does warning on exhaustive switch statement (enum)

### ISSUE THREAD

**Issue by AshTheMash** [pmd/pmd#2606]
[java]

**Affects PMD Version:** 
6.24.0

**Description:**
I programmed an enumeration and used it in a switch clause without default case. PMD gives the warning "Switch statements should have a default value" about this. I also implemented in my code the test scenario used in PMD on exhaustive switch statements. Also there using the tool I get a warning which is quite odd. So there is a slight chance that I miss a property on the rule or similar. However, I spent hours on this and could not find a solution.

**Exception Stacktrace:**
None.

**Code Sample demonstrating the issue:**
**GradeSystem.java**

    package myPackage;

    public enum GradeSystem {
        SysA, SysB
    }

**Foo.java**

    package myPackage;

    public class Foo {

        public void foo(GradeSystem system) {
            switch(system) {
            case SysA:
                break;
            case SysB:
                break;
            }
        }
    }

**Steps to reproduce:**

Please provide detailed steps for how we can reproduce the bug.

1.  Create GradeSystem.java and Foo.java files from above
2.  Use pmd.bat -l java -no-cache -dir [the directory with the files] -R pmd_gkm_ruleset.xml -f html

**Running PMD through:**
[CLI ]

**Comment by oowekyala** [comment]
Sorry for the late feedback. Normally this should be handled if you provide a full classpath. PMD uses class files so that we know in the `Foo.java` file what are the enum constants of `GradeSystem`. Here we need that info to determine whether the switch is exhaustive.

In this case you should provide a path to compiled classes with the [`-auxclasspath`](https://pmd.github.io/latest/pmd_userdocs_cli_reference.html#auxclasspath) CLI option, something like `-auxclasspath myartifact.jar`

I don't think, this is a bug, so I'm going to close this. Please reopens if you get this warning when the auxclasspath is correctly configured

**Comment by Searen1984** [comment]
Hey thanks for the answer. I tried using the -auxclasspath parameter. However, then I get a lot of exceptions. I found this issue which might be related [#2738](https://github.com/pmd/pmd/issues/2738). So I was updating the used PMD binary to v6.29. However, this did not help. Here is one of these exceptions:
Nov. 21, 2020 7:18:16 NACHM. net.sourceforge.pmd.lang.rule.AbstractRuleChainVisitor visitAll
WARNING: Exception applying rule SwitchStmtsShouldHaveDefault on file D:\ScholarCertificate\de.gkm.common\src\main\java\de\gkm\common\db\elements\Grade1To6.java, continuing with next rule
java.lang.RuntimeException: 
                //SwitchStatement[@DefaultCase = false() and @ExhaustiveEnumSwitch = false()]
             had problem: The string "" cannot be cast to a boolean
	at net.sourceforge.pmd.lang.rule.xpath.SaxonXPathRuleQuery.evaluate(SaxonXPathRuleQuery.java:145)
	at net.sourceforge.pmd.lang.rule.XPathRule.evaluate(XPathRule.java:176)
	at net.sourceforge.pmd.lang.rule.XPathRule.apply(XPathRule.java:158)
	at net.sourceforge.pmd.lang.internal.DefaultRulechainVisitor.visit(DefaultRulechainVisitor.java:26)
	at net.sourceforge.pmd.lang.rule.AbstractRuleChainVisitor.visitAll(AbstractRuleChainVisitor.java:99)
	at net.sourceforge.pmd.RuleChain.apply(RuleChain.java:72)
	at net.sourceforge.pmd.RuleSets.apply(RuleSets.java:145)
	at net.sourceforge.pmd.SourceCodeProcessor.processSource(SourceCodeProcessor.java:194)
	at net.sourceforge.pmd.SourceCodeProcessor.processSourceCodeWithoutCache(SourceCodeProcessor.java:107)
	at net.sourceforge.pmd.SourceCodeProcessor.processSourceCode(SourceCodeProcessor.java:89)
	at net.sourceforge.pmd.SourceCodeProcessor.processSourceCode(SourceCodeProcessor.java:51)
	at net.sourceforge.pmd.processor.PmdRunnable.call(PmdRunnable.java:78)
	at net.sourceforge.pmd.processor.PmdRunnable.call(PmdRunnable.java:24)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: ValidationException: The string "" cannot be cast to a boolean
	at net.sf.saxon.type.ValidationFailure.makeException(ValidationFailure.java:233)
	at net.sf.saxon.type.ValidationFailure.asAtomic(ValidationFailure.java:254)
	at net.sf.saxon.expr.GeneralComparison.compare(GeneralComparison.java:607)
	at net.sf.saxon.expr.SingletonComparison.effectiveBooleanValue(SingletonComparison.java:102)
	at net.sf.saxon.expr.QuantifiedExpression.effectiveBooleanValue(QuantifiedExpression.java:226)
	at net.sf.saxon.expr.BooleanExpression.effectiveBooleanValue(BooleanExpression.java:182)
	at net.sf.saxon.expr.FilterIterator$NonNumeric.matches(FilterIterator.java:182)
	at net.sf.saxon.expr.FilterIterator.getNextMatchingItem(FilterIterator.java:65)
	at net.sf.saxon.expr.FilterIterator.next(FilterIterator.java:44)
	at net.sourceforge.pmd.lang.rule.xpath.SaxonXPathRuleQuery.evaluate(SaxonXPathRuleQuery.java:127)
	... 18 more

java.lang.reflect.InvocationTargetException
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at net.sourceforge.pmd.lang.ast.xpath.Attribute.getValue(Attribute.java:98)
	at net.sourceforge.pmd.lang.ast.xpath.saxon.AttributeNode.atomize(AttributeNode.java:60)
	at net.sf.saxon.om.SingleNodeIterator.atomize(SingleNodeIterator.java:129)
	at net.sf.saxon.expr.AxisAtomizingIterator.next(AxisAtomizingIterator.java:43)
	at net.sf.saxon.expr.QuantifiedExpression.effectiveBooleanValue(QuantifiedExpression.java:221)
	at net.sf.saxon.expr.BooleanExpression.effectiveBooleanValue(BooleanExpression.java:182)
	at net.sf.saxon.expr.FilterIterator$NonNumeric.matches(FilterIterator.java:182)
	at net.sf.saxon.expr.FilterIterator.getNextMatchingItem(FilterIterator.java:65)
	at net.sf.saxon.expr.FilterIterator.next(FilterIterator.java:44)
	at net.sourceforge.pmd.lang.rule.xpath.SaxonXPathRuleQuery.evaluate(SaxonXPathRuleQuery.java:127)
	at net.sourceforge.pmd.lang.rule.XPathRule.evaluate(XPathRule.java:176)
	at net.sourceforge.pmd.lang.rule.XPathRule.apply(XPathRule.java:158)
	at net.sourceforge.pmd.lang.internal.DefaultRulechainVisitor.visit(DefaultRulechainVisitor.java:26)
	at net.sourceforge.pmd.lang.rule.AbstractRuleChainVisitor.visitAll(AbstractRuleChainVisitor.java:99)
	at net.sourceforge.pmd.RuleChain.apply(RuleChain.java:72)
	at net.sourceforge.pmd.RuleSets.apply(RuleSets.java:145)
	at net.sourceforge.pmd.SourceCodeProcessor.processSource(SourceCodeProcessor.java:194)
	at net.sourceforge.pmd.SourceCodeProcessor.processSourceCodeWithoutCache(SourceCodeProcessor.java:107)
	at net.sourceforge.pmd.SourceCodeProcessor.processSourceCode(SourceCodeProcessor.java:89)
	at net.sourceforge.pmd.SourceCodeProcessor.processSourceCode(SourceCodeProcessor.java:51)
	at net.sourceforge.pmd.processor.PmdRunnable.call(PmdRunnable.java:78)
	at net.sourceforge.pmd.processor.PmdRunnable.call(PmdRunnable.java:24)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: java.lang.ExceptionInInitializerError
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at java.base/java.lang.Class.getEnumConstantsShared(Class.java:3553)
	at java.base/java.lang.Class.getEnumConstants(Class.java:3529)
	at org.apache.commons.lang3.EnumUtils.getEnumMap(EnumUtils.java:59)
	at net.sourceforge.pmd.lang.java.ast.ASTSwitchStatement.isExhaustiveEnumSwitch(ASTSwitchStatement.java:83)
	... 32 more
Caused by: java.lang.UnsupportedOperationException: No class provided, and an appropriate one cannot be found.
	at org.apache.logging.log4j.LogManager.callerClass(LogManager.java:576)
	at org.apache.logging.log4j.LogManager.getLogger(LogManager.java:601)
	at org.apache.logging.log4j.LogManager.getLogger(LogManager.java:588)
	at de.gkm.common.db.elements.GradeAToD.<clinit>(GradeAToD.java:16)
	... 40 more
java.lang.reflect.InvocationTargetException
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at net.sourceforge.pmd.lang.ast.xpath.Attribute.getValue(Attribute.java:98)
	at net.sourceforge.pmd.lang.ast.xpath.saxon.AttributeNode.atomize(AttributeNode.java:64)
	at net.sf.saxon.om.SingleNodeIterator.atomize(SingleNodeIterator.java:129)
	at net.sf.saxon.expr.AxisAtomizingIterator.next(AxisAtomizingIterator.java:43)
	at net.sf.saxon.expr.QuantifiedExpression.effectiveBooleanValue(QuantifiedExpression.java:221)
	at net.sf.saxon.expr.BooleanExpression.effectiveBooleanValue(BooleanExpression.java:182)
	at net.sf.saxon.expr.FilterIterator$NonNumeric.matches(FilterIterator.java:182)
	at net.sf.saxon.expr.FilterIterator.getNextMatchingItem(FilterIterator.java:65)
	at net.sf.saxon.expr.FilterIterator.next(FilterIterator.java:44)
	at net.sourceforge.pmd.lang.rule.xpath.SaxonXPathRuleQuery.evaluate(SaxonXPathRuleQuery.java:127)
	at net.sourceforge.pmd.lang.rule.XPathRule.evaluate(XPathRule.java:176)
	at net.sourceforge.pmd.lang.rule.XPathRule.apply(XPathRule.java:158)
	at net.sourceforge.pmd.lang.internal.DefaultRulechainVisitor.visit(DefaultRulechainVisitor.java:26)
	at net.sourceforge.pmd.lang.rule.AbstractRuleChainVisitor.visitAll(AbstractRuleChainVisitor.java:99)
	at net.sourceforge.pmd.RuleChain.apply(RuleChain.java:72)
	at net.sourceforge.pmd.RuleSets.apply(RuleSets.java:145)
	at net.sourceforge.pmd.SourceCodeProcessor.processSource(SourceCodeProcessor.java:194)
	at net.sourceforge.pmd.SourceCodeProcessor.processSourceCodeWithoutCache(SourceCodeProcessor.java:107)
	at net.sourceforge.pmd.SourceCodeProcessor.processSourceCode(SourceCodeProcessor.java:89)
	at net.sourceforge.pmd.SourceCodeProcessor.processSourceCode(SourceCodeProcessor.java:51)
	at net.sourceforge.pmd.processor.PmdRunnable.call(PmdRunnable.java:78)
	at net.sourceforge.pmd.processor.PmdRunnable.call(PmdRunnable.java:24)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: java.lang.NoClassDefFoundError: Could not initialize class de.gkm.common.db.elements.GradeAToD
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at java.base/java.lang.Class.getEnumConstantsShared(Class.java:3553)
	at java.base/java.lang.Class.getEnumConstants(Class.java:3529)
	at org.apache.commons.lang3.EnumUtils.getEnumMap(EnumUtils.java:59)
	at net.sourceforge.pmd.lang.java.ast.ASTSwitchStatement.isExhaustiveEnumSwitch(ASTSwitchStatement.java:83)
	... 32 more

**Comment by adangel** [comment]
@Searen1984 That seems to be a different issue. How to you call PMD? What's the content of your ruleset?

**Comment by adangel** [comment]
Ok, I've seen now the exception: "NoClassDefFoundError: Could not initialize class de.gkm.common.db.elements.GradeAToD" - it seems, you auxclasspath is not complete...

**Comment by Searen1984** [comment]
@adangel Thanks for the quick reply. Here is my call to PMD:
"[...]\pmd-bin-6.29.0\bin\pmd.bat" -l java -auxclasspath [...]\classes -no-cache -dir [...] -R pmd_gkm_ruleset.xml -f html -ignorelist [...]\pmd_ignore.txt > "[...]\PMD_Result.html"


[pmd_ruleset.zip](https://github.com/pmd/pmd/files/5578221/pmd_ruleset.zip)

**Comment by adangel** [comment]
I guess, you enum looks something like this:

```java
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

public enum GradeAToD {
    SysA, SysB;

   private static final Logger LOG = LogManager.getLogger();
}
```

Try to replace `LogManager.getLogger()` with `LogManager.getLogger(GradeAToD.class)`.

The way PMD works with auxclasspath/type resolution is, that it actually loads the classes into the VM. That means static initializers like static fields/methods are executed. For some reason, log4j can't determine the classname by magic in that case, so giving the logger name directly should help.

It might also be an indication, that your auxclasspath is not complete. Usually, you have log4j as a jar file, you can put this complete jar file (and any other dependency, that it might need) onto the auxclasspath, e.g.
`-auxclasspath [...]\classes;lib\log4j.jar;lib\otherLib.jar`

**Comment by Searen1984** [comment]
Thanks that lets me execute PMD with the -auxclasspath switch. However, my initial problem remains and I got another problem now. I still get the warning "Switch statements should have a default value" and there are additionally some false positives "Avoid unused private methods such as...". I checked some of them and they seem to be static methods.

**Comment by adangel** [comment]
Have a look at the example/sourcecode here: https://github.com/adangel/scratchpad/tree/master/pmd-issue-2606 and compare this to your source code.

**Comment by Searen1984** [comment]
Thanks for your feedback. I really appreciate your support. 

Your example runs as expected. But please try my example. It does not run as expected and is reproducable. Here is what to do:
Under Windows (as I do not use Linux):

Create **only** this class under src\myPackage (the ```this``` statement is important):
```
package myPackage;

public enum SomeEnum {
  ENUM1, ENUM2;

  public boolean isValid() {
    switch (this) {
      case ENUM1:
      case ENUM2:
        return true;
    }
    
    return false;
  }
}
```

Clean: rd /s "target"
Compile (here was a mistake the "myPackage" is necessary): javac -d target src\myPackage\*.java
Run PMD: pmd.bat -d src -auxclasspath target -R pmd_ruleset.xml -no-cache -f text
--> one violation: [...]\src\myPackage\SomeEnum.java:7:  Switch statements should have a default label

Run PMD without auxclasspath: pmd.bat -d src -R pmd_ruleset.xml -no-cache -f text
--> one violation: [...]\src\myPackage\SomeEnum.java:7:  Switch statements should have a default label

**Comment by Searen1984** [comment]
I could reduce many of the false positives using the auxclasspath. However, I still get some messages of the PMD tool. I could reduce it to the example given above and the following. Please add the following class to the same package as described in my last message:
```
package myPackage;

import java.util.List;

public class Foo {
    
  private SomeEnum e;
  
  public SomeEnum getE() {
      return e;
  }
    
  public void setValue(List<SomeEnum> list) {
    list.forEach(eIn -> {
        switch (eIn) {
          case ENUM1:
            this.e = SomeEnum.ENUM2;
            break;
          case ENUM2:
            this.e = SomeEnum.ENUM1;
            break;
        }
      });
  }
}
```

I used the same commands as given in the last message and then I get with those two results:
src\myPackage\Foo.java:15:      Switch statements should have a default label
src\myPackage\SomeEnum.java:7:  Switch statements should have a default label

Attached you find the project I am using.
[test.zip](https://github.com/pmd/pmd/files/5583195/test.zip)

**Comment by adangel** [comment]
Thanks for sharing your code. Your examples are indeed a false positive, hence I've created a new issue: #2946 
For the other with lambdas, we already had a similar issue: #2743
